runtime/lua/vim/filetype/detect.lua

```diff
@@ -787,9 +862,67 @@ function M.scd(bufnr)
   end
 end
 
-function M.sh(path, bufnr) end
+-- Also called from filetype.lua
+function M.sh(path, bufnr, name)
+  if vim.fn.did_filetype() or path:find(vim.g.ft_ignore_pat) then
+    -- Filetype was already detected or detection should be skipped
+    return
+  end
 
-function M.shell(path, bufnr) end
+  if matchregex(name, [[\c\<csh\>]]) then
+    -- Some .sh scripts contain #!/bin/csh.
+    M.shell(path, bufnr, 'csh')
+    return
+    -- Some .sh scripts contain #!/bin/tcsh.
+  elseif matchregex(name, [[\c\<tcsh\>]]) then
+    M.shell(path, bufnr, 'tcsh')
+    return
+    -- Some .sh scripts contain #!/bin/zsh.
+  elseif matchregex(name, [[\c\<zsh\>]]) then
+    M.shell(path, bufnr, 'zsh')
+    return
+  elseif matchregex(name, [[\c\<ksh\>]]) then
+    vim.b[bufnr].is_kornshell = 1
+    vim.b[bufnr].is_bash = nil
+    vim.b[bufnr].is_sh = nil
+  elseif
+    (vim.g.bash_is_sh and vim.g.bash_is_sh ~= 0)
+    or matchregex(name, [[\c\<bash\>]])
+    or matchregex(name, [[\c\<bash2\>]])
+  then
+    vim.b[bufnr].is_bash = 1
+    vim.b[bufnr].is_kornshell = nil
+    vim.b[bufnr].is_sh = nil
+  elseif matchregex(name, [[\c\<sh\>]]) then
+    vim.b[bufnr].is_sh = 1
+    vim.b[bufnr].is_kornshell = nil
+    vim.b[bufnr].is_bash = nil
+  end
+  M.shell(path, bufnr, 'sh')
+end
+
+-- For shell-like file types, check for an "exec" command hidden in a comment, as used for Tcl.
+-- Also called from scripts.vim, thus can't be local to this script. [TODO]
+function M.shell(path, bufnr, name)
+  if vim.fn.did_filetype() or matchregex(path, vim.g.ft_ignore_pat) then
+    -- Filetype was already detected or detection should be skipped
+    return
+  end
+  local prev_line = ''
+  for _, line in ipairs(getlines(bufnr, 2, -1)) do
+    line = line:lower()
+    if line:find('%s*exec%s') and not prev_line:find('^%s*#.*\\$') then
+      -- Found an "exec" line after a comment with continuation
+      local n = line:gsub('%s*exec%s+([^ ]*/)?', '', 1)
+      if matchregex(n, [[\c\<tclsh\|\<wish]]) then
+        vim.bo[bufnr].filetype = 'tcl'
+        return
+      end
+    end
+    prev_line = line
+  end
+  vim.bo[bufnr].filetype = name
```

Comment:
Note that all of these functions must return a non-nil value if they set the filetype (we actually *really* want them to just return the filetype rather than setting it directly, but we can do that in a follow up).